impl/src/main/java/org/ehcache/internal/store/disk/OffHeapDiskStore.java

```diff
@@ -0,0 +1,375 @@
+        DiskWriteThreadPool writeWorkers = new DiskWriteThreadPool("identifier", config.getConcurrency());
+
+        Factory<FileBackedStorageEngine<K, OffHeapValueHolder<V>>> storageEngineFactory = FileBackedStorageEngine.createFactory(source, config
+                .getSegmentDataPageSize(), keyPortability, elementPortability, writeWorkers, false);
+
+        EhcachePersistentSegmentFactory<K, OffHeapValueHolder<V>> factory = new EhcachePersistentSegmentFactory<K, OffHeapValueHolder<V>>(
+            source,
+            storageEngineFactory,
+            config.getInitialSegmentTableSize(),
+            evictionVeto,
+            mapEvictionListener, false);
+            EhcachePersistentConcurrentOffHeapClockCache m = new EhcachePersistentConcurrentOffHeapClockCache<K, OffHeapValueHolder<V>>(input, factory);
+
+
+
+
+        m.bootstrap(input);
+        return m;
+      } catch (IOException e) {
+        source.close();
+        throw e;
+      }
+    } catch (Exception e) {
+      LOGGER.info("Index file was corrupt. Deleting data file " + dataFile.getAbsolutePath() +". " + e.getMessage());
+      LOGGER.debug("Exception during recovery", e);
+      return createBackingMap(size, keySerializer, valueSerializer, evictionVeto);
+    } finally {
+      fin.close();
+    }
+  }
+  
+  private EhcachePersistentConcurrentOffHeapClockCache<K, OffHeapValueHolder<V>> createBackingMap(long size, Serializer<K> keySerializer, Serializer<V> valueSerializer, Predicate<Map.Entry<K, OffHeapValueHolder<V>>> evictionVeto) {
+    HeuristicConfiguration config = new HeuristicConfiguration(size);
+    MappedPageSource source;
+    try {
+      source = new MappedPageSource(fileBasedPersistenceContext.getDataFile(), size);
+    } catch (IOException e) {
+      // TODO proper exception
+      throw new RuntimeException(e);
+    }
+    PersistentPortability<K> keyPortability = persistent(new SerializerPortability<K>(keySerializer));
+    PersistentPortability<OffHeapValueHolder<V>> elementPortability = persistent(new OffHeapValueHolderPortability<V>(valueSerializer));
+    DiskWriteThreadPool writeWorkers = new DiskWriteThreadPool("identifier", config.getConcurrency());
+
+    Factory<FileBackedStorageEngine<K, OffHeapValueHolder<V>>> storageEngineFactory = FileBackedStorageEngine.createFactory(source, config
+        .getSegmentDataPageSize(), keyPortability, elementPortability, writeWorkers, true);
+
+    EhcachePersistentSegmentFactory<K, OffHeapValueHolder<V>> factory = new EhcachePersistentSegmentFactory<K, OffHeapValueHolder<V>>(
+        source,
+        storageEngineFactory,
+        config.getInitialSegmentTableSize(),
+        evictionVeto,
+        mapEvictionListener, true);
+    return new EhcachePersistentConcurrentOffHeapClockCache<K, OffHeapValueHolder<V>>(factory, config.getConcurrency());
+
+  }
+
+  private static <K, V> Predicate<Map.Entry<K, OffHeapValueHolder<V>>> wrap(EvictionVeto<? super K, ? super V> delegate) {
```

Comment:
This method is identical to `OffHeapStore.wrap` and so can probably be extracted to the `AbstractOffHeapStore`